src/lib.rs

```diff
@@ -578,12 +578,55 @@ fn readline_raw<H: Helper>(
     user_input
 }
 
-fn readline_direct() -> Result<String> {
-    let mut line = String::new();
-    if io::stdin().read_line(&mut line)? > 0 {
-        Ok(line)
-    } else {
-        Err(error::ReadlineError::Eof)
+fn readline_direct(validator: &Option<impl Validator>) -> Result<String> {
+    let mut input = String::new();
+
+    loop {
+        match io::stdin().read_line(&mut input)? {
+            0 => return Err(error::ReadlineError::Eof),
+            _ => {
+                // Remove trailing newline
+                if input.ends_with('\n') {
+                    input.pop();
+                    if input.ends_with('\r') {
+                        input.pop();
+                    }
+                }
+
+                match validator.as_ref() {
+                    None => return Ok(input),
+                    Some(v) => {
+                        struct SimpleInvoke<'a> {
+                            s: &'a str,
+                        }
+
+                        impl<'a> keymap::Invoke for SimpleInvoke<'a> {
+                            fn input(&self) -> &str {
+                                self.s
+                            }
+                        }
+
+                        let mut ctx = SimpleInvoke { s: &input };
+                        let mut ctx = validate::ValidationContext::new(&mut ctx);
+
+                        match v.validate(&mut ctx)? {
+                            validate::ValidationResult::Incomplete => {
+                                // Add a newline back onto the end of the buffer
+                                input.push('\n');
+                            }
+                            validate::ValidationResult::Invalid(_msg) => {
+                                // TODO: present the msg back to stdout?
+                                return Ok(input);
+                            }
+                            validate::ValidationResult::Valid(_msg) => {
+                                // TODO: present the msg back to stdout?
+                                return Ok(input);
```

Comment:
I am not entirely sure how to handle the case where `Valid` or `Invalid` return a message to the user.

Also `Invalid` maybe shouldn't return the input just yet :thinking:

Reply:
We may find an answer on how to handle message in [python-prompt-toolkit](https://python-prompt-toolkit.readthedocs.io/en/master/pages/asking_for_input.html#input-validation).

Reply:
It looks like it ignores the message as well when not in TTY mode:

```bash
➜ echo -e 'hello\b\b\b\b\b45' | python test.py
Warning: Input is not a terminal (fd=0).
^[[65;1RGive a number: 45
You said: 45
```

Reply:
Nevermind, I forgot that I need to "press Enter" before erasing the string "hello". It crashes, if I try that :stuck_out_tongue_closed_eyes:

Reply:
```bash
➜ echo -e 'hello\n\b\b\b\b\b45' | python test.py
Warning: Input is not a terminal (fd=0).
^[[65;1RGive a number: 45hello
Traceback (most recent call last):
  File "test.py", line 20, in <module>
    number = int(prompt('Give a number: ', validator=NumberValidator()))
  File "/home/basile/dev/temp/venv/lib/python3.8/site-packages/prompt_toolkit/shortcuts/prompt.py", line 1397, in prompt
    return session.prompt(
  File "/home/basile/dev/temp/venv/lib/python3.8/site-packages/prompt_toolkit/shortcuts/prompt.py", line 1020, in prompt
    return self.app.run(
  File "/home/basile/dev/temp/venv/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 887, in run
    return loop.run_until_complete(
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/basile/dev/temp/venv/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 816, in run_async
    return await _run_async2()
  File "/home/basile/dev/temp/venv/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 792, in _run_async2
    result = await _run_async()
  File "/home/basile/dev/temp/venv/lib/python3.8/site-packages/prompt_toolkit/application/application.py", line 733, in _run_async
    result = await f
EOFError
```

Reply:
Ok, I believe this is now properly handled.